servers/src/main/java/tachyon/master/next/IndexedSet.java

```diff
@@ -0,0 +1,231 @@
+   * o.equals(o2)). If this set already contains the object, the call leaves the set unchanged.
+   *
+   * @param object the object to add
+   * @return true if the object is successfully added to all indexes, otherwise false
+   */
+  public boolean add(T object) {
+    synchronized (mLock) {
+      boolean success = mObjects.add(object);
+      for (FieldIndex<T> field : mIndexMap.keySet()) {
+        Map<Object, Set<T>> fieldValueToSet = mSetIndexedByFieldValue.get(mIndexMap.get(field));
+        Object value = field.getFieldValue(object);
+        if (fieldValueToSet.containsKey(value)) {
+          success = success && fieldValueToSet.get(value).add(object);
+        } else {
+          fieldValueToSet.put(value, Sets.newHashSet(object));
+        }
+      }
+      return success;
+    }
+  }
+
+  /**
+   * Return the set of all objects, the returned set is backed up by the internal set, any changes
+   * to the internal set are reflected in the returned set, and vice-versa.
+   *
+   * @return a set of all objects
+   */
+  // TODO(cc) Let this class implement Iterable to avoid this method, all use cases are to foreach
+  // this set.
+  public Set<T> all() {
+    synchronized (mLock) {
+      return mObjects;
+    }
+  }
+
+  /**
+   * Whether there is an object with the specified field value in the set.
+   *
+   * @param index the field index
+   * @param value the field value
+   * @return true if there is one such object, otherwise false
+   */
+  public boolean contains(FieldIndex<T> index, Object value) {
+    synchronized (mLock) {
+      return getByFieldInternal(index, value) != null;
+    }
+  }
+
+  /**
+   * Get a subset of objects with the specified field value. O(n) in worst case due to the copy of
+   * the internal set.
+   *
+   * @param index the field index
+   * @param value the field value to be satisfied
+   * @return the set of objects or an empty set if no such object exists
+   */
+  public Set<T> getByField(FieldIndex<T> index, Object value) {
+    synchronized (mLock) {
+      Set<T> set = getByFieldInternal(index, value);
+      return set == null ? new HashSet<T>() : Sets.newHashSet(set);
+    }
+  }
+
+  /**
+   * Get the first object from the set of objects with the specified field value.
+   *
+   * @param index the field index
+   * @param value the field value
+   * @return the object or null if there is no such object
+   */
+  public T getFirstByField(FieldIndex<T> index, Object value) {
+    synchronized (mLock) {
+      Set<T> all = getByFieldInternal(index, value);
+      return all == null ? null : all.iterator().next();
+    }
+  }
+
+  /**
+   * Remove an object from the set.
+   *
+   * @param object the object to remove
+   * @return true if the object is in the set and removed successfully, otherwise false
+   */
+  public boolean remove(T object) {
+    synchronized (mLock) {
+      boolean success = mObjects.remove(object);
+      for (FieldIndex<T> field : mIndexMap.keySet()) {
```

Comment:
You eventually use the key and value of the mIndexMap. Maybe you should just iterate over the entrySet().